### PR TITLE
ENS proposal lifecycle

### PR DESCRIPTION
**Pull request by mcgingras**
Proposal lifecycle management for ENS

### CODE REVIEW

File: src/app/proposals/draft/components/stages/SubmitForm.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  ProposalChecklist,
+} from "@prisma/client";
+import { useContractRead, useAccount, useBlockNumber } from "wagmi";
+import { ENSGovernorABI } from "@/lib/contracts/abis/ENSGovernor";
+import Tenant from "@/lib/tenant/tenant";
+import RequestSponsorshipForm from "../RequestSponsorshipForm";
+import { useForm, FormProvider } from "react-hook-form";
+import SponsorActions from "../../../sponsor/components/SponsorActions";
+
+// TODO: either read from contract or add to tenant
```
Comment (mcgingras) [inline src/app/proposals/draft/components/stages/SubmitForm.tsx:15]: @mcgingras address

File: src/app/proposals/draft/components/RequestSponsorshipForm.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+          });
+          setIsPending(false);
+          if (!res.ok) {
+            // toast?
+          } else {
+            // pretty funky way to revalidate path?
+            // I'm curious what this is actually doing internally
+            // better than a full page refresh though...
+            // wonder if there is a way to optimistically update data
+            invalidatePath(draftProposal.id);
```
Comment (andreitr) [inline src/app/proposals/draft/components/RequestSponsorshipForm.tsx:81]: I used similar approach on staking and it seemed to work pretty well. I like how you extracted the method into an action.

File: src/app/proposals/draft/components/SimulateStatusPill.tsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  simulationId,
+}: {
+  status: string;
+  simulationId: string | null;
+}) => {
+  return (
+    <span
+      className={cn(
+        "relative border rounded-lg p-2",
+        status === "UNCONFIRMED" &&
```
Comment (andreitr) [inline src/app/proposals/draft/components/SimulateStatusPill.tsx:16]: In the future, let's extract status into a constant.

File: src/app/proposals/draft/components/SocialProposalForm.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  }, [proposalType]);
+
+  return (
+    <div className="space-y-6">
+      <div>
+        <h3 className="text-stone-900 font-semibold">
+          Voting strategy and choices
+        </h3>
+        <p className="mt-2 stone-700">
+          Choose the voting stratgy and options for your proposal.
```
Comment (andreitr) [inline src/app/proposals/draft/components/SocialProposalForm.tsx:54]: typo in the word strategy

File: src/app/proposals/draft/components/TransferTransactionForm.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        functionName: "transfer",
+        args: [recipient as `0x${string}`, BigInt(parseUnits(amount, 18))],
+      });
+
+      setValue(`transactions.${index}.calldata`, calldata);
+      // TODO: replace with tenant governor address
+      // right now it's ENS test address
+      setValue(
+        `transactions.${index}.target`,
+        "0xca83e6932cf4F03cDd6238be0fFcF2fe97854f67"
```
Comment (andreitr) [inline src/app/proposals/draft/components/TransferTransactionForm.tsx:65]: Yep, let's replace this with the tenant governor

File: src/app/proposals/draft/components/TransferTransactionForm.tsx [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  const recipient = watch(`transactions.${index}.recipient`);
+  const amount = watch(`transactions.${index}.amount`);
+
+  useEffect(() => {
+    if (recipient && amount && isAddress(recipient)) {
+      // calc the calldata for transfer
+      const calldata = encodeFunctionData({
+        abi: transferABI,
+        functionName: "transfer",
+        args: [recipient as `0x${string}`, BigInt(parseUnits(amount, 18))],
```
Comment (andreitr) [inline src/app/proposals/draft/components/TransferTransactionForm.tsx:57]: Let's get token decimals from the tenant as well

File: src/app/proposals/draft/components/dialogs/SponsorOnchainProposalDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  return (
+    <VStack alignItems="items-center">
+      <VStack className="w-full bg-white rounded-xl">
+        <VStack>
+          <VStack className="w-full">
+            <img
+              src={
+                isLoading
+                  ? `/images/action-pending.svg`
+                  : `/images/ens_success.svg`
```
Comment (andreitr) [inline src/app/proposals/draft/components/dialogs/SponsorOnchainProposalDialog.tsx:33]: The asset images are now in the tenant.ui. You can get them 

const {ui} = Tenant.current()
ui.assets.success
ui.assets.pending

File: src/app/proposals/draft/components/dialogs/SponsorOnchainProposalDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                It might take up to a minute for the changes to be reflected.
+              </div>
+            )}
+            <div>
+              <UpdatedButton
+                fullWidth={true}
+                type="primary"
+                isLoading={isLoading}
+                onClick={async () => {
+                  // TODO: redirect to the proposal page once we have indexing available
```
Comment (andreitr) [inline src/app/proposals/draft/components/dialogs/SponsorOnchainProposalDialog.tsx:52]: Can you expand on this?

File: src/app/proposals/draft/components/stages/DraftForm.tsx [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+  ProposalTypeMetadata,
+} from "../../types";
+import {
+  ProposalDraft,
+  ProposalSocialOption,
+  ProposalDraftTransaction,
+} from "@prisma/client";
+import ExecutableProposalForm from "../ExecutableProposalForm";
+import SocialProposalForm from "../SocialProposalForm";
+// import FileInput from "../form/FileInput";
```
Comment (andreitr) [inline src/app/proposals/draft/components/stages/DraftForm.tsx:27]: Comment

File: src/app/proposals/draft/components/stages/DraftForm.tsx [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+            </div>
+          </FormCard.Section>
+          <FormCard.Section>
+            {proposalType === ProposalType.EXECUTABLE ? (
+              <ExecutableProposalForm />
+            ) : (
+              <SocialProposalForm />
+            )}
+          </FormCard.Section>
+          {/* <FormCard.Section>
```
Comment (andreitr) [inline src/app/proposals/draft/components/stages/DraftForm.tsx:153]: Comment

File: src/app/proposals/draft/components/stages/GithubPRForm.tsx [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+      await createGithubChecklistItem({
+        draftProposalId: draftProposal.id,
+        creatorAddress: address,
+        link: "",
+      });
+
+      setIsSkipPending(false);
+      router.push(`/proposals/draft/${draftProposal.id}?stage=3`);
+    } catch (e) {
+      console.error(e);
```
Comment (andreitr) [inline src/app/proposals/draft/components/stages/GithubPRForm.tsx:56]: Do we need this console here?

File: src/app/proposals/draft/components/stages/GithubPRForm.tsx [hunk 12]
```diff
@@ -0,0 +1,170 @@
+"use client";
+
+import { useAccount } from "wagmi";
+import { useState } from "react";
+import { z } from "zod";
+// import Tenant from "@/lib/tenant/tenant";
```
Comment (andreitr) [inline src/app/proposals/draft/components/stages/GithubPRForm.tsx:6]: Dead import

File: src/app/proposals/draft/components/stages/SubmitForm.tsx [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+const THRESHOLD = 100000000000000000000000;
```
Comment (andreitr) [inline src/app/proposals/draft/components/stages/SubmitForm.tsx:16]: This constant is used in several places. Let's ether load it from a contract or move it into constants.

File: src/app/proposals/draft/components/stages/TempCheckForm.tsx [hunk 14]
```diff
@@ -0,0 +1,121 @@
+"use client";
+
+import { useState } from "react";
+import { z } from "zod";
+// import Tenant from "@/lib/tenant/tenant";
```
Comment (andreitr) [inline src/app/proposals/draft/components/stages/TempCheckForm.tsx:5]: Dead import

File: src/app/proposals/draft/components/stages/TempCheckForm.tsx [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+      });
+      router.push(`/proposals/draft/${draftProposal.id}?stage=1`);
+    } catch (e) {
+      console.error(e);
+    }
+  };
+
+  return (
+    <form
+    //   action={async (formData: FormData) => {
```
Comment (andreitr) [inline src/app/proposals/draft/components/stages/TempCheckForm.tsx:64]: comment

File: src/app/proposals/draft/components/stages/TempCheckForm.tsx [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    //     await tempCheckAction(formData);
+    //     // this should probably be order in the tenent list + 1
+    //     redirect(`/proposals/draft?stage=1`);
+    //   }}
+    >
+      <FormCard>
+        <FormCard.Section>
+          <div className="w-full rounded-md h-[350px] block relative">
+            <Image
+              src="/images/temp_check.png"
```
Comment (andreitr) [inline src/app/proposals/draft/components/stages/TempCheckForm.tsx:74]: This needs to be a tenant-specific image... ens_temp_check or something like this. Just a note, no action right now.

File: src/app/proposals/draft/schemas/tempCheckSchema.ts [hunk 17]
```diff
@@ -0,0 +1,17 @@
+import { z } from "zod";
+
+const isURL = (value: string) => {
```
Comment (andreitr) [inline src/app/proposals/draft/schemas/tempCheckSchema.ts:3]: move into utils?

File: src/app/proposals/draft/utils/createSnapshot.ts [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+
+export async function createSnapshot({
+  address,
+  proposal,
+}: {
+  address: string | null;
+  proposal: ProposalDraft;
+}) {
+  // TODO: replace with tenent to get proper chainId
+  const walletClient = getWalletClient(11155111);
```
Comment (andreitr) [inline src/app/proposals/draft/utils/createSnapshot.ts:67]: Chain ids can be pulled from the tenant contracts

File: src/app/proposals/draft/utils/createSnapshot.ts [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+  const publicClient = getPublicClient(1);
+
+  if (!address) {
+    throw new Error("address not available");
+  }
+
+  const description =
+    proposal.abstract +
+    "\n" +
+    // `${
```
Comment (andreitr) [inline src/app/proposals/draft/utils/createSnapshot.ts:77]: comment?

File: src/app/proposals/draft/utils/createSnapshot.ts [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+    proposal.abstract;
+
+  const blockNumber = await publicClient.getBlockNumber();
+  const timestamp = Math.floor(new Date().getTime() / 1000);
+  const message: SnapshotProposalMessage & { [key: string]: unknown } = {
+    from: address!,
+    space:
+      process.env.REACT_APP_DEPLOY_ENV === "prod"
+        ? "ens.eth"
+        : process.env.TESTNET_SNAPSHOT_SPACE ?? "michaelagora.eth",
```
Comment (andreitr) [inline src/app/proposals/draft/utils/createSnapshot.ts:95]: michaelagora.eth ?

File: src/app/proposals/draft/utils/getInputData.ts [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+): {
+  inputData: BasicInputData;
+} {
+  const description =
+    "# " +
+    proposal.title +
+    "\n\n" +
+    proposal.abstract +
+    "\n" +
+    // `${
```
Comment (andreitr) [inline src/app/proposals/draft/utils/getInputData.ts:17]: comment?

File: src/app/proposals/draft/utils/getInputData.ts [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+      calldatas.push("0x");
+    } else {
+      proposal.transactions.forEach((t) => {
+        targets.push(ethers.getAddress(t.target) as `0x${string}`);
+        values.push(ethers.parseEther(t.value.toString() || "0"));
+        calldatas.push(t.calldata as `0x${string}`);
+      });
+    }
+  } catch (e) {
+    console.error(e);
```
Comment (andreitr) [inline src/app/proposals/draft/utils/getInputData.ts:48]: Do we need to handle the error state here?

File: src/app/proposals/draft/utils/github.ts [hunk 23]
```diff
@@ -0,0 +1,351 @@
+"use server";
```
Comment (andreitr) [inline src/app/proposals/draft/utils/github.ts:1]: A lot of todos, comments, and console.logs. Let's nuke the code we are not using.

File: src/app/proposals/sponsor/components/SponsorActions.tsx [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+              setIsSnapshotPending(true);
+              const proposalId = await createSnapshot({
+                address: address as `0x${string}`,
+                proposal: draftProposal,
+              });
+
+              const snapshotLink = isProd
+                ? `https://snapshot.org/#/ens.eth/proposal/${proposalId}`
+                : `https://testnet.snapshot.org/#/michaelagora.eth/proposal/${proposalId}`;
+
```
Comment (andreitr) [inline src/app/proposals/sponsor/components/SponsorActions.tsx:69]: Is there a reason we are hardcoding things to michaelagora.eth?

File: src/app/proposals/sponsor/components/SponsorActions.tsx [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+                type: "SPONSOR_ONCHAIN_DRAFT_PROPOSAL",
+                params: {
+                  redirectUrl: "/",
+                  txHash: data?.hash as `0x${string}`,
+                },
+              });
+            }
+          } catch (error) {
+            console.error(error);
+            // toast?
```
Comment (andreitr) [inline src/app/proposals/sponsor/components/SponsorActions.tsx:97]: toast?

File: src/components/Proposals/ProposalsList/actions/createProposalDraft.tsx [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+      abstract: "",
+      audit_url: "",
+      author_address: address,
+      sponsor_address: "",
+      stage: firstStage.stage,
+      // TODO: need a way to generalize this to the current tenant
+      dao: {
+        connectOrCreate: {
+          where: {
+            dao_slug: DaoSlug.ENS,
```
Comment (andreitr) [inline src/components/Proposals/ProposalsList/actions/createProposalDraft.tsx:24]: The Dao Slug can be pulled from the tenant

File: src/lib/tenant/tenant.ts [hunk 27]
```diff
@@ -53,6 +53,13 @@ export default class Tenant {
     return this._token;
   }
 
+  public get blockExplorer(): string {
```
Comment (andreitr) [inline src/lib/tenant/tenant.ts:56]: We had similar thoughts on how to handle block explorer links. I moved this stuff out into utils but the logic is very similar.

**Review by andreitr — COMMENTED** [review]
What a massive PR! Left a bunch a comments - mostly nits.

File: src/app/proposals/draft/components/stages/TempCheckForm.tsx (hunk 16 above)
Comment (mcgingras) [reply src/app/proposals/draft/components/stages/TempCheckForm.tsx:74]: rename

**Review by andreitr — APPROVED** [review]
LGTM

File: src/app/page.jsx [hunk 28]
```diff
@@ -8,6 +12,8 @@ import NeedsMyVoteProposalsList from "@/components/Proposals/NeedsMyVoteProposal
 import ProposalsList from "@/components/Proposals/ProposalsList/ProposalsList";
 import { proposalsFilterOptions } from "@/lib/constants";
 import Tenant from "@/lib/tenant/tenant";
+import MyDraftProposals from "@/components/Proposals/DraftProposals/MyDraftProposals";
+import MySponsorshipRequests from "@/components/Proposals/DraftProposals/MySponsorshipRequests";
 
 // Revalidate cache every 60 seconds
 export const revalidate = 60;
```
[the remark this answers is not in the document]
Comment (mcgingras) [reply src/app/page.jsx:19]: I don't think I added this revalidate export

File: src/app/proposals/draft/utils/github.ts (hunk 23 above)
Comment (mcgingras) [reply src/app/proposals/draft/utils/github.ts:1]: I think there are actually mistakes in this file that I need to go back and fix. Thanks for calling this out.

**Pull request merged**